### ISSUE TITLE
[Catalog utility functions] find_chunking_info

### ISSUE THREAD

**Issue by Thomas-Moore-Creative** [ACCESS-NRI/access-nri-intake-catalog#218]
# Is your feature request related to a problem? Please describe.

To enable a better understanding of the underlying NetCDF data structure so settings like `xarray_open_kwargs` can be used effectively requires discovery of the native file chunking.

# Describe the feature you'd like

1. a place for the community to help build utility functions that would support the https://github.com/ACCESS-NRI/access-nri-intake-catalog
2. a specific function to "find the native chunking information" for a dataset in the catalog that first-time users have the ability to use and understand

# Describe alternatives you've considered

Writing my own `pre-alpha` functions here: https://github.com/Thomas-Moore-Creative/ACDtools/blob/main/ACDtools like `find_chunking_info` but I'd like a place to collaborate on utilities that was easier for the whole community to see and share.

# Additional context

- I'm a user of the data.
- I'm not trained professionally as a software engineer, but I'm working on it.
- I'm employing the `from tabulate import tabulate` function to make simple tables display in the `Jupyter` UI.

**Comment by aidanheerdegen** [comment]
Hola @Thomas-Moore-Creative,

This sounds cool. I'm thinking the best place for something like this might be a repo on the ACCESS Community Hub organisation. It makes it more straightforward to collaborate as we could make you an admin on the repo.

https://github.com/ACCESS-Community-Hub

Then we could point to it from this repo

Does that sound like a good way forward?

**Comment by Thomas-Moore-Creative** [comment]
Sounds fine to me @aidanheerdegen - thanks.  Do you, @rbeucher, @dougiesquire, or any of your software engineering gurus have advice on how to  structure this repo so it's portable, flexible, and available to all on NCI?

**Comment by aidanheerdegen** [comment]
Dougie is on leave, so he's out of the picture. 

To make it available on `gadi` I'd say we should add `conda` packaging. We could also arrange to publish it to the [accessnri anaconda channel](https://anaconda.org/accessnri), or create another access community channel.

We can deal with that later. 

As for repo structure, first decision might be [flat layout vs src layout](https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/), and then isolate functionality in sub-directories.

Is that the sort of thing you were thinking about @Thomas-Moore-Creative?

Do you have any opinions @marc-white?

**Comment by marc-white** [comment]
I think the main thing to determine is a question of scope. What _exactly_ are you trying to do? Is it just doing some stuff to work out the native chunking of netCDF files, or are you looking to expand this to include more tools later down the track?

Then, once you've worked out the answer to that question, that will inform your answer to the next question: should this come in as a part of `access-nri-intake-catalog`, or should it be spun off into its own utility package?

**Comment by Thomas-Moore-Creative** [comment]
Thanks @marc-white.  

What I'm trying to do is get my projects done, which requires using the `access-nri-intake-catalog`, and for me that means data discovery, building search filters, and understanding data structure to allow optimal analysis-ready-data workflows to be built for specific datasets.  

I highlighted just one type of very simple utility that I'm building ( "find the native chunking information" ) in this issue but I am wondering out loud if there is a better place to be developing helper utilities than my personal repos?  Maybe the questions are:

- does ACCESS-NRI think utilities to help users discover and load data from the `access-nri-intake-catalog` is a real and general need for the community?
- where and how ( in what repo ) might ACCESS-NRI want these functions to be collaborated on, if at all?

**Comment by charles-turner-1** [comment]
> As for repo structure, first decision might be [flat layout vs src layout](https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/), and then isolate functionality in sub-directories.

I'd recommend going with src layout for consistency - it seems to be Dougie's preferred layout, and would keep things consistent with [this package](https://github.com/ACCESS-NRI/access-nri-intake-catalog) and the related [intake-dataframe-catalog](https://github.com/ACCESS-NRI/intake-dataframe-catalog).

As to whether this should be included within `access-nri-intake-catalog` or as a standalone package, I would suggest the latter. Lots of the functionality of the catalog, eg. loading datasets, is actually performed by `intake-esm`, and I suspect that this might cause complications. My vote would be for a separate package - something like `access-intake-utils` - and then we try to keep the interdependencies as minimal as possible.

**Comment by Thomas-Moore-Creative** [comment]
> My vote would be for a separate package - something like `access-intake-utils` - and then we try to keep the interdependencies as minimal as possible.

From a users point of view this makes sense to me.  Thanks for the advice.

Can we start with an `access-intake-utils` repo in https://github.com/ACCESS-Community-Hub, as suggested by @aidanheerdegen above?

**Comment by rbeucher** [comment]
I agree with @ charles-turner-1, A separate package is the way to go for now. Feel free to start in ACCESS-Community-Hub

**Comment by charles-turner-1** [comment]
I've added a new `get_disk_chunks` function over at https://github.com/ACCESS-NRI/access-intake-utils/commit/d3a284e62558bb9fbeda2ec9d9aba92cfae1b97e.

Now I've clocked what `find_chunking_info` is doing underneath, I think we might be able to build out some functionality based on that which might make it easier for users to plan out their chunking strategy.

With that said, xarray does use disk chunk sizes when reading files from disk (see https://github.com/pydata/xarray/pull/7948), so using
```python
.to_dask(
    xarray_open_kwargs = {'chunks' : 'auto'},
)
```

will take these disk chunks into account by default, and I think we should generally recommend to users - I've done so over at cosima-recipes. 

What `'chunks':'auto'` can't do is select chunks based on the subsequent operations the user wants to do - I think we can probably address that, at least to an extent.

**Comment by Thomas-Moore-Creative** [comment]
Thanks @charles-turner-1 

As I've said before my code is both very opinionated ( for my needs ) and also possibly not the most solid software engineering on the planet.

I note and agree with your comments re: `'chunks':'auto'`.  To be clear my POV was that often you'd want to increase the size of chunks based on multiples of the native NetCDF chunking that a function like `get_disk_chunks` would expose to the user inside the workflow.

**Comment by Thomas-Moore-Creative** [comment]
> Presumably there's a large fraction of the community who are routinely scratching their heads

My plan is to continue this discussion with the community in that July "training" talk.  When I get around to planning that I'll touch base with you and others to get your perspective.

**Comment by charles-turner-1** [comment]
I've added functionality [here](https://github.com/ACCESS-NRI/access-intake-utils/pull/2/files) which should address this chunking issue. It's not fully ready to go yet, but the TLDR; is that you should be able to use this to do something like
```python
>>> import intake
>>> import access_intake_utils

>>> cat = intake.cat.access_nri
>>> expt_datastore = cat[expt_str_name]

>>> dataset = expt_datastore.search(variable='xyz')

# Now we set chunking kwargs.
# Let's assume lat & lon have disk chunks of 60
>>> chunks = {'time': 10, 'lat' : 300,  'lon' : 250,  'time' : -1}

# Validate chunks
>>> validated_chunks =  access_intake_utils.validate_chunkspec(dataset, chunks)
ChunkingWarning: Specified chunks are not integer multiples of the disk chunks. Returning suggested chunks as a dictionary.

>>> print(validated_chunks)
{'time': 10, 'lat' : 300,  'lon' : 300,  'time' : -1}

>>> dataset.to_dask(xarray_open_kwargs={'chunks' : validated_chunks}
```

Once I get it cleaned up a bit further, merged & released I'll link & close this issue.

**Comment by charles-turner-1** [comment]
Okay, I've released the package with the chunking validation [here](https://github.com/ACCESS-NRI/access-intake-utils/blob/main/examples/validate_chunkspec.ipynb).

It's not in the conda analysis3 environments yet, but should be accessible through `pip install access-intake-utils`. Conda release to come.

**Comment by rbeucher** [comment]
Great!

**Comment by charles-turner-1** [comment]
Merged and should now be available in conda/analysis3-25.05!
